### ISSUE TITLE
How to get metrics for specific time serie from global model

### ISSUE THREAD

**Issue by mariosantosprivate** [awslabs/gluon-ts#1143]
Hello, I'm having some trouble when trying to compare the Deep AR and Prophet models.

I'm training the Deep AR model on roughly 1000 time series (daily sales per store data), I've selected a couple of time series and trained fitted them using Prophet. From my understanding, Deep AR creates a global model from all the time series.

With the aggregate metrics I can get the metrics across all time series and time steps from the Deep AR model (I'm assuming an average of each metric for each store in this case), however is there a way to extract the metrics for a specific time series?

For example, I've fitted the Prophet model with the time series from store "XYZ" and get an RMSE value, Deep AR was fitted on 1000 time series and gives me an aggregate RMSE based on all time series, is there a way to extract the metrics specifically for store "XYZ" so I can compare them to Prophet's? 

Originally I thought that it would be the item_metric but it seems that returns the metrics of each time step aggregated across all time series (so having time series with x time steps will result in item_metrics of x rows), what I needed would be an item_metrics of n length where n is the number of time series, giving me an average metric for each ts.

**Comment by PascalIversen** [comment]
Hey @mariosantosprivate,

if I understand you correctly, the `item_metrics` are what you want. See this simple example:
```python


import mxnet as mx

from gluonts.dataset.repository.datasets import get_dataset
from gluonts.model.simple_feedforward import SimpleFeedForwardEstimator
from gluonts.trainer import Trainer

dataset = get_dataset("m4_hourly", regenerate=False)

estimator = SimpleFeedForwardEstimator(freq="H", prediction_length=24, trainer=Trainer(epochs=1, ctx=mx.cpu()))

predictor = estimator.train(dataset.train, num_workers=None)
from gluonts.evaluation.backtest import make_evaluation_predictions
forecast_it, ts_it = make_evaluation_predictions(
    dataset=dataset.test,  # test dataset
    predictor=predictor,  # predictor
    num_samples=100,  # number of sample paths we want for evaluation
)

from gluonts.evaluation import Evaluator
evaluator = Evaluator(quantiles=[0.1, 0.5, 0.9])
ts = iter(list(ts_it))
fc = iter(list(forecast_it))
agg_metrics, item_metrics = evaluator(ts,fc , num_series=len(dataset.test))
item_metrics
```
![Screenshot 2020-11-11 at 17 22 42](https://user-images.githubusercontent.com/66594767/98836936-9c151180-2442-11eb-9ed3-35b5ed3a53c7.png)

We have one metric value for each time series in the test data (for you the item_id would correspond to the store).

**Comment by mariosantosprivate** [comment]
> Hey @mariosantosprivate,
> 
> if I understand you correctly, the `item_metrics` are what you want. See this simple example:
> 
> ```python
> import mxnet as mx
> 
> from gluonts.dataset.repository.datasets import get_dataset
> from gluonts.model.simple_feedforward import SimpleFeedForwardEstimator
> from gluonts.trainer import Trainer
> 
> dataset = get_dataset("m4_hourly", regenerate=False)
> 
> estimator = SimpleFeedForwardEstimator(freq="H", prediction_length=24, trainer=Trainer(epochs=1, ctx=mx.cpu()))
> 
> predictor = estimator.train(dataset.train, num_workers=None)
> from gluonts.evaluation.backtest import make_evaluation_predictions
> forecast_it, ts_it = make_evaluation_predictions(
>     dataset=dataset.test,  # test dataset
>     predictor=predictor,  # predictor
>     num_samples=100,  # number of sample paths we want for evaluation
> )
> 
> from gluonts.evaluation import Evaluator
> evaluator = Evaluator(quantiles=[0.1, 0.5, 0.9])
> ts = iter(list(ts_it))
> fc = iter(list(forecast_it))
> agg_metrics, item_metrics = evaluator(ts,fc , num_series=len(dataset.test))
> item_metrics
> ```
> 
> ![Screenshot 2020-11-11 at 17 22 42](https://user-images.githubusercontent.com/66594767/98836936-9c151180-2442-11eb-9ed3-35b5ed3a53c7.png)
> 
> We have one metric value for each time series in the test data (for you the item_id would correspond to the store).

@PascalIversen Hi, yes it seems that it should be however I am not getting the expected return from item_metrics, this is definitely related to this question I just asked (if you think it shouldn't be a separate question I'll delete it and place it here!)

https://github.com/awslabs/gluon-ts/issues/1144

**Comment by PascalIversen** [comment]
I think your confusion might be the following: DeepAR  is a global but not a multivariate model. Please have a look at #1095. The length of your dataset has to be the number of time series in your dataset.
The targets of each time series should have shape (t), where t is the length of the time-series.
Draw some inspiration from this code on how to create an appropriate dataset for univariate models like DeepAR.

```python
import pandas as pd
import numpy as np
from gluonts.dataset.common import ListDataset

indices = pd.date_range(start = "2020-01-01", end = "2020-10-30")
#creating some random data
target = np.random.randint(0, 100, size=(3,len(indices))) + np.linspace((1,10,50),(5,20,100), len(indices)).T.astype("int")

time_series_dicts = []
for time_series in target:
    time_series_dicts.append({"target": time_series, "start": indices[0]})
dataset = ListDataset(time_series_dicts, freq=indices[0].freq)
```
Let me know if you experience problems.

**Comment by mariosantosprivate** [comment]
> I think your confusion might be the following: DeepAR is a global but not a multivariate model. Please have a look at #1095. The length of your dataset has to be the number of time series in your dataset.
> The targets of each time series should have shape (t), where t is the length of the time-series.
> Draw some inspiration from this code on how to create an appropriate dataset for univariate models like DeepAR.
> 
> ```python
> import pandas as pd
> import numpy as np
> from gluonts.dataset.common import ListDataset
> 
> indices = pd.date_range(start = "2020-01-01", end = "2020-10-30")
> #creating some random data
> target = np.random.randint(0, 100, size=(3,len(indices))) + np.linspace((1,10,50),(5,20,100), len(indices)).T.astype("int")
> 
> time_series_dicts = []
> for time_series in target:
>     time_series_dicts.append({"target": time_series, "start": indices[0]})
> dataset = ListDataset(time_series_dicts, freq=indices[0].freq)
> ```
> 
> Let me know if you experience problems.

It is the case that the target of each time series has shape(t), since the test_target_values_shape is (number_of_stores, number_of_time_steps), each store holds an array of the target values for every day (t)

To clarify, this is how I'm creating the list data sets:

```python
train_ds = ListDataset([
    {
        FieldName.TARGET: target,
        FieldName.START: start,
        FieldName.FEAT_DYNAMIC_REAL: fdr,
        FieldName.FEAT_STATIC_CAT: fsc
    }
    for (target, start, fdr, fsc) in zip(train_target_values,
                                         dates,
                                         train_dynamic_features_list,
                                         stat_cat)
], freq="D")

test_ds = ListDataset([
    {
        FieldName.TARGET: target,
        FieldName.START: start,
        FieldName.FEAT_DYNAMIC_REAL: fdr,
        FieldName.FEAT_STATIC_CAT: fsc
    }
    for (target, start, fdr, fsc) in zip(test_target_values,
                                         dates,
                                         test_dynamic_features_list,
                                         stat_cat)
], freq="D")
```

Where:
test_target_values shape - (934, 894) 934 time series (stores) each of 894 time steps (days)
train_target_values shape (excluding last 48 time steps which is my forecast horizon) - (934, 846) 
test_dynamic_features shape - (934, 29, 894) time series each with 29 dynamic real features for 894 time steps
train dynamic features shape (excluding last 48 time steps) - (934, 29, 846) 934 
stat_cat shape - (934 ,4) 934 time series each with 4 static categoricals

This was also the approach used in this notebook from a kaggle competition:
https://www.kaggle.com/steverab/m5-forecasting-competition-gluonts-template

Isn't this the right approach?

**Comment by PascalIversen** [comment]
Ok, that seems good to me. Can you try to write a reproducible example?

The following code works fine, try to see whats the difference between your code and this example:

```python
import mxnet as mx
from gluonts.dataset.repository.datasets import get_dataset
from gluonts.model.simple_feedforward import SimpleFeedForwardEstimator
from gluonts.trainer import Trainer

import pandas as pd
import numpy as np
from gluonts.dataset.common import ListDataset

indices = pd.date_range(start = "2020-01-01", end = "2020-10-30", freq="H")
#creating some random data
target = np.random.randint(0, 100, size=(3,len(indices))) + np.linspace((1,10,50),(5,20,100), len(indices)).T.astype("int")

time_series_dicts = []
for time_series in target:
    time_series_dicts.append({"target": time_series, "start": indices[0]})
dataset = ListDataset(time_series_dicts, freq=indices[0].freq)


estimator = SimpleFeedForwardEstimator(freq="H", prediction_length=2, trainer=Trainer(epochs=1, ctx=mx.cpu()))

predictor = estimator.train(dataset, num_workers=None)
from gluonts.evaluation.backtest import make_evaluation_predictions
forecast_it, ts_it = make_evaluation_predictions(
    dataset=dataset,  # test dataset
    predictor=predictor,  # predictor
    num_samples=100,  # number of sample paths we want for evaluation
)

from gluonts.evaluation import Evaluator
evaluator = Evaluator(quantiles=[0.1, 0.5, 0.9], num_workers=None)
ts = list(ts_it)
fc = list(forecast_it)
agg_metrics, item_metrics = evaluator(ts,fc , num_series=len(dataset))
item_metrics
```

**Comment by mariosantosprivate** [comment]
@PascalIversen Ok so the problem was my dates array, which had a length of 894 (the time series length, as I didn't think to include duplicates), so it was just a silly mistake on my end. Thank you very much for the speedy assistance!
